File: backend/writeBuffer.js
// Write-back batch buffer: aggregates search counts in memory and flushes to
// Postgres on size (batchSizeN) or interval (flushIntervalMs). A crash loses at
// most one un-flushed window — acceptable for approximate counts (DESIGN.md).
const config = require('./config');
const metrics = require('./metrics');

class WriteBuffer {
  constructor(flushHandler) {
    this.buf = new Map();
    this.handler = flushHandler;
    this.timer = null;
    this.flushing = false;
  }

  add(query) {
    query = query.toLowerCase().trim();
    if (!query) return;
    metrics.m.searchesReceived++;
    this.buf.set(query, (this.buf.get(query) || 0) + 1);
    if (this.buf.size >= config.batchSizeN) this.flush();
  }

  async flush() {
    if (this.flushing || this.buf.size === 0) return;
    this.flushing = true;
    const window = this.buf;     // swap out the current window...
    this.buf = new Map();        // ...start a fresh one (no await between)
    try {
      await this.handler(window);
    } catch (e) {
      // don't let a transient flush error crash the process (window is lost,
      // same as the documented crash trade-off)
      console.error('[flush] error:', e.message);
    } finally {
      this.flushing = false;
    }
  }

  start() { this.timer = setInterval(() => this.flush(), config.flushIntervalMs); }

  async stop() {
    if (this.timer) clearInterval(this.timer);
    await this.flush();          // final drain
  }

  pending() { return this.buf.size; }
}

module.exports = { WriteBuffer };

File: backend/write_buffer.py
"""Write-back batch buffer: aggregates search counts in memory and flushes to
Postgres on size (BATCH_SIZE_N) or interval (FLUSH_INTERVAL_T). A crash loses at
most one un-flushed window — acceptable for approximate counts (DESIGN.md §8)."""
import asyncio
from typing import Awaitable, Callable, Dict

from . import config, metrics

FlushHandler = Callable[[Dict[str, int]], Awaitable[None]]


class WriteBuffer:
    def __init__(self, flush_handler: FlushHandler):
        self._buf: Dict[str, int] = {}
        self._handler = flush_handler
        self._size_event = asyncio.Event()
        self._task: asyncio.Task | None = None
        self._stopping = False

    def add(self, query: str) -> None:
        query = query.lower().strip()
        if not query:
            return
        metrics.searches_received += 1
        self._buf[query] = self._buf.get(query, 0) + 1
        if len(self._buf) >= config.BATCH_SIZE_N:
            self._size_event.set()  # wake the loop for a size-triggered flush

    async def _flush_once(self) -> int:
        if not self._buf:
            return 0
        window = self._buf            # swap out the current window...
        self._buf = {}                # ...and start a fresh one (no await between)
        self._size_event.clear()
        await self._handler(window)
        return len(window)

    async def _loop(self) -> None:
        while not self._stopping:
            try:
                await asyncio.wait_for(self._size_event.wait(), timeout=config.FLUSH_INTERVAL_T)
            except asyncio.TimeoutError:
                pass  # time-based flush
            await self._flush_once()

    def start(self) -> None:
        self._task = asyncio.create_task(self._loop())

    async def stop(self) -> None:
        self._stopping = True
        self._size_event.set()
        if self._task:
            await self._task
        await self._flush_once()  # final drain so nothing is left buffered

    def pending(self) -> int:
        return len(self._buf)
